### PR TITLE
Propagate EffectiveScope to parameters of anonymous delegate types.

### LINKED ISSUE #63078: Remove or reduce use of `ParameterSymbol.DeclaredScope`
Consider if we remove `ParameterSymbol.DeclaredScope` and use `EffectiveScope` only, or at least reduce the use of `DeclaredScope` perhaps to source parameters only.

_Based on discussion from https://github.com/dotnet/roslyn/pull/62783#discussion_r932747637_

### LINKED ISSUE #63565: Implicitly typed lambdas with `[UnscopedRef]` don't match their own anonymous delegate types
**Version Used**: 
```
commit e041ae9802b68f7c6f6903ec92b56d27ff0caec3
Author: Youssef Victor <youssefvictor00@gmail.com>
Date:   Wed Aug 24 16:00:43 2022 +0200

    Delete extra character in Syntax.xml (#63561)
```

**Steps to Reproduce**:

```cs

using System.Diagnostics.CodeAnalysis;

public ref struct RefStruct { }
public struct RegularStruct { }

delegate void D1([UnscopedRef] ref RefStruct s);
delegate void D2([UnscopedRef] out RegularStruct s);

public class C
{
  public void M()
  {
    var x = ([UnscopedRef] ref RefStruct s) => { }; // error CS8986: The 'scoped' modifier of parameter 's' doesn't match target '<anonymous delegate>'.
    var y = ([UnscopedRef] out RegularStruct s) => { }; // error CS8986: The 'scoped' modifier of parameter 's' doesn't match target '<anonymous delegate>'.
    D1 q = ([UnscopedRef] ref RefStruct s) => { };
    D2 w = ([UnscopedRef] out RegularStruct s) => { };
  }
}

namespace System.Diagnostics.CodeAnalysis
{
  [AttributeUsage(AttributeTargets.All, Inherited = false)]
  internal sealed class UnscopedRefAttribute : Attribute
  {
    public UnscopedRefAttribute()
    {
    }
  }
}
```

**Expected Behavior**:
No errors for lambda expressions assigned to `var` variables. The anonymous delegates created for them should have `[UnscopedRef]` annotations on their parameters

**Actual Behavior**:
Lambda expressions with completely valid signatures (as demonstrated by `D1`, `D2` and corresponding lambda expressions) produce errors due to mismatches with their own implicit types

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #63565.
Closes #63078.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @RikkiGibson, @dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceParameterSymbol.cs [hunk 1]
```diff
@@ -222,10 +222,21 @@ public sealed override RefKind RefKind
             }
         }
 
-        internal sealed override DeclarationScope DeclaredScope => _scope;
+        /// <summary>
+        /// The declared scope. From source, this is from the <c>scope</c> keyword only.
+        /// </summary>
+        internal DeclarationScope DeclaredScope => _scope;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceParameterSymbol.cs:228]: >DeclaredScope [](http://example.com/codeflow?start=34&length=13)

It looks like this fixes [https://github.com/dotnet/roslyn/issues/63078](https://github.com/dotnet/roslyn/issues/63078).

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs [hunk 2]
```diff
@@ -492,7 +492,7 @@ private TypeWithAnnotations ComputeType(Binder binder, SyntaxNode syntax, Bindin
                 allowRefOrOut: false,
                 allowThis: false,
                 addRefReadOnlyModifier: addRefReadOnlyModifier,
-                diagnostics: diagnostics);
+                diagnostics: diagnostics).Cast<SourceParameterSymbol, ParameterSymbol>();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs:495]: >.Cast<SourceParameterSymbol, ParameterSymbol>() [](http://example.com/codeflow?start=41&length=47)

We're casting to `ParameterSymbol` in many callers of `MakeParameters()`. Where are we using the change to `SourceParameterSymbol`, and would it make sense to cast that case to `SourceParameterSymbol` instead?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 3]
```diff
@@ -16609,5 +16609,157 @@ public struct Int32 { }
...
+    var b = ([UnscopedRef] out RegularStruct s) => { };
+    var c = ([UnscopedRef] out RefStruct s) => { };
+
+    D1 x = ([UnscopedRef] ref RefStruct s) => { };
+    D2 y = ([UnscopedRef] out RegularStruct s) => { };
+    D3 z = ([UnscopedRef] out RefStruct s) => { };
+  }
+}
+";
+            var comp = CreateCompilation(new[] { source, UnscopedRefAttributeDefinition }, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.ReleaseDll);
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:16642]: >targetFramework: TargetFramework.NetCoreApp [](http://example.com/codeflow?start=91&length=43)

Minor: Is `targetFramework` needed?

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceParameterSymbol.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceParameterSymbol.cs:228]: > It looks like this fixes #63078.

Thanks. I'll mark the issue as closed by this PR.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs:495]: > We're casting to `ParameterSymbol` in many callers of `MakeParameters()`. Where are we using the change to `SourceParameterSymbol`, and would it make sense to cast that case to `SourceParameterSymbol` instead?

Is there a specific concern? This is a safe in-place cast. I intentionally changed return type to a more specific type in order to ensure that assumptions made in ```ParameterHelpers.ReportParameterErrors``` are justified.

**Comment by AlekseyTs** [comment]
@RikkiGibson, @dotnet/roslyn-compiler For the second review.

**Comment by AlekseyTs** [comment]
@RikkiGibson, @dotnet/roslyn-compiler For the second review.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 4]
```diff
@@ -613,35 +613,34 @@ private static TypeWithAnnotations DelegateReturnTypeWithAnnotations(MethodSymbo
...
-                var lambdaSymbol = new LambdaSymbol(
-                    Binder,
-                    Binder.Compilation,
-                    Binder.ContainingMemberOrLambda,
-                    _unboundLambda,
-                    parameterTypes,
-                    parameterRefKinds,
-                    refKind: default,
-                    returnType: default);
+                lambdaSymbol = createLambdaSymbol();
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:643]: It seems strange that we will sometimes create a lambda symbol which gets thrown away in order to produce information like the inferred return type or effective scope of parameters, and then later on presumably produce a "final" lambda symbol. Is this a sign that there's some design debt here which could be addressed in the future?

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:643]: > Is this a sign that there's some design debt here which could be addressed in the future?

I do not think so. As long as this doesn't become a source of performance problem, that isn't something that I would worry about.

**Comment by jcouv** [comment]
@AlekseyTs I can't find this behavior described in the spec. Should we document it? Did we confirm with LDM this is desirable?

**Comment by AlekseyTs** [comment]
> I can't find this behavior described in the spec. Should we document it? Did we confirm with LDM this is desirable?

My understanding is that that we were propagating something already that didn't make sense in many scenarios. Triage determined that the behavior was not desirable and was classified as a bug. At this point I do not remember whether this behavior or the previous behavior was explicitly documented in the spec. If you believe we should have a different behavior, feel free to open a bug or bring this for LDM discussion.

**Comment by jcouv** [comment]
Discussed with @cston and filed https://github.com/dotnet/roslyn/issues/64670 for follow-up (spec needs updating)

**Pull request merged**